dot/rpc/subscription/websocket.go

```diff
@@ -213,6 +213,8 @@ func (c *WSConn) initStorageChangeListener(reqID float64, params interface{}) (L
 }
 
 func (c *WSConn) initBlockListener(reqID float64, _ interface{}) (Listener, error) {
+	// TODO (ed) here the imported Channel get unregistered and closed in http.Stop
+	//  there doesn't seem to be an un-subscribe for this
```

Comment:
where does `BlockListener.Channel` get closed? i don't see it in the `BlockListener.Stop`

Reply:
Can we move this code to a constructor for `BlockListener`.

Reply:
It seems there is logic to close imported channel in dot/http.go  line 194-5, which in inside HTTPServer.Stop.  There is also within Listener.Stop a call to cancelWithTimeout that gets called when unsubscribe is called.  Should we move the imported channel close into Listener.Stop, then call listener.Stop from within HTTPServer.Stop?

Reply:
Yea we should localise closing channels (managing internal state) to receiver functions of the struct.

Reply:
@timwu20 the `BlockListener.Channel` channel is closed when `l.wsconn.BlockAPI.UnregisterImportedChannel(l.ChanID)` is called, we should not close the `BlockListener.Channel` directly because the `BlockAPI` is the owner of the channel. Every listener has a `Stop` method that calls the function `cancelWithTimeout` which will stop any running listener's goroutines, which in turn will call the `defer function` to unregister any active channel

Reply:
But, the implementation of `Unregister...` function is to just remove the channel from the map without closing it, and that is the problem. @edwardmack I think we could close the imported channel inside the `defer` function of each `Listen` method to keep things closer.

Reply:
I added imported channel close to defer in Listen methods.